README.md
ffi-magics++
============
[homepage](https://github.com/manoute/ffi-magicsplus)

DESCRIPTION
-----------

Ruby bindings for magics++ using FFI.
[Magics++](http://www.ecmwf.int/products/data/software/magics++.html) is the latest generation of the [ECMWF](http://www.ecmwf.int)'s Meteorological plotting software 

FEATURES
--------

**Close to Magics++ C api :** just remove mag_ from C functions : 

    // With C api
    mag_open  
    mag_coast 
    ...

    # With Ruby Api
    MagPlus.open
    Magplus.coast
    ...

**Except :**

  - mag_new :

        // C api
        mag_new(superpage)
        mag_new(page)
        ...
        
        # Ruby api
        MagPlus.new_page
        MagPlus.new_super_page  
        ...

  - When setting array, there is no need to provide dimensions :

        // C api
        mag_set1i("foo",array,12) 
        
        # Ruby api
        MagPlus.set1i("foo",array) 

  - After close, all parameters are reset to their default values, to avoid this behaviour, use close(false) or open(false) do ... end.

** Blocks can be given to open, enabling automatic close :**

        MagPlus.open do |m|
          m.setc('output_filename','foo')
          ...
          m.coast
        end 

** Setting parameter can be done with = :**

        Magplus.subpage_upper_right_latitude = 30.0

** A hash and block can be given to methods :**
        
        # C-api like
        Magplus.setr('subpage_upper_right_latitude',30.0)
        Magplus.setr('subpage_upper_right_longitude',30.0)
        Magplus.new_subpage
        
        # Same thing with a hash
        Magplus.new_subpage({:subpage_upper_right_latitude => 30.0,
          :subpage_upper_right_longitude => 30.0})

        # Same thing with a block
        Magplus.new_subpage do |c|
          c.setr('subpage_upper_right_latitude',30.0)
          c.subpage_upper_right_longitude = 30.0
        end

** params hash is filled with the settings :**

        MagPlus.open
        MagPlus.subpage_lower_left_latitude = 30.0
        MagPlus.params #  {:subpage_lower_left_latitude => 30.0}
        MagPlus.reset_all # reset all parameters to their default value
        MagPlus.params #  {}
    
** Works with ruby array or narray :**
    
        # With Ruby array
        MagPlus.set1r("foo",[44.0, 51.0,52.0,53.0])

        # With NArray
        MagPlus.set1r("foo",NArray[44.0, 51.0,52.0,53.0])

PROBLEMS
--------

Experimental.

Only tested with Debian package and custom package on Archlinux.

Bufr not tested...working?

EXAMPLES
--------

Plotting an array of data

    require 'rubygems' # optional
    require 'ffi-magics++'

    MagPlus.open do |c|
      c.setc("output_fullname",'2D.ps')

      # Block syntax
      c.new_subpage do |sub|
        sub.subpage_lower_left_latitude = 40.0
        sub.subpage_lower_left_longitude = -10.0
        sub.subpage_upper_right_latitude = 55.0
        sub.subpage_upper_right_longitude = 10.0
      end

      # C-api syntax
      c.setr("input_field_initial_longitude",-180.0)
      c.setr("input_field_longitude_step",1.0)
      c.setr("input_field_initial_latitude",-90.0)
      c.setr("input_field_latitude_step",1.0)
      data = []
      (0..180).each {|i| data << (360.times.inject([]){|r,e| r << e})}
      c.set2r('input_field', data)

      # Hash syntax
      c.cont({:contour => 'off', :contour_label => 'off', 
        :contour_shade => 'on', :contour_shade_method => 'area_fill',
        :contour_level_count => 30})

      c.coast
    end


Using module as a mixing and plotting a grib file

    require 'rubygems' # optional
    require 'ffi-magics++'

    class Toto
      include MagPlus

      def grib_plot
        open do
          self.output_fullname = 'foo.pdf'

          grib({:grib_input_type => 'file',
                :grib_input_file_name => 'foo.grb'})

          cont({:contour => 'on', :contour_shade => 'on',
                :contour_shade_method => 'area_fill'})
          
          coast
        end
      end
    end

    Toto.new.grib_plot



For other examples, see examples directory.

REQUIREMENTS
------------
* A ruby implementation ([RVM](http://rvm.beginrescueend.com) allows to easily install, manage and work with multiple ruby environments.)
* Magics++ must be installed :

      apt-get install magics++ 

      or have a look at Magics++ installation on their homepage.

* Need 'ffi'.
* 'narray' is optionnal.
* 'rspec' for testing. 
* 'yard' to generate the doc.

INSTALL
-------

    [sudo] gem install ffi-magicsplus 

 
RUNNING SPECS/TESTS
-------------------

From the root source tree, type :

    [RUBYLIB=.:lib] rake test:all 

LICENSE
-------

(The MIT License)

Copyright (c) 2010 Mathieu Deslandes

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
Version 0.2.2
-------------

- params hash filled with paremeters provided to Magics++
- parameters reet to their default value just before close

Version 0.2.1
-------------

- a block and a hash can be given to most methods
- setting parameter with =

Version 0.1.1
-------------

initial release




examples/wavewatch3.rb
require 'ffi-magics++'
require 'open-uri'
require 'date'

class WW3
  include MagPlus

  def host
    'ftpprd.ncep.noaa.gov'
  end
  
  def path
    'pub/data/nccf/com/wave/prod/wave.' +
    Date.today.strftime('%Y%m%d') + 
    '/multi_1.glo_30m.t00z.f000.grib2'
  end 

  def url 
    URI::FTP.build({:host => host, :path => path})
  end
  
  def fetch_grib_file
    puts "Downloading Wave Watch III grib file (today, 00h)...\n" +
      "See http://polar.ncep.noaa.gov/waves/"
    
    data = url.read
    File.open('ww3.grib2','w') {|f| f.write data} 
    puts "File downloaded and saved to ww3.grib2"
  end
  
  def plot_grib_file
    open do 
      # Output
      setc("output_fullname",'ww3.pdf')
      setc("output_format","pdf")

      # Projection
      setc('subpage_map_projection','polar_stereographic')

      # Grid and coast
      setr("subpage_lower_left_latitude",20.0)
      setr("subpage_lower_left_longitude",-40.0)
      setr("subpage_upper_right_latitude",70.0)
      setr("subpage_upper_right_longitude",10.0)
      setr("map_grid_latitude_reference",0.0)
      setr("map_grid_longitude_reference",0.0)
      setr("map_grid_longitude_increment",5.0)
      setr("map_grid_latitude_increment",5.0)
      setc("map_grid_colour","black")
      setc("map_grid_line_style","dash")
      setc("map_coastline_land_shade", "on")
      setc("map_coastline_land_shade_colour", "grey")

      # Grib file and field selection
      setc("grib_input_type","file")
      setc("grib_input_file_name","ww3.grib2")
      seti("grib_field_position",5)
      grib

      # Contour
      setc("contour","on")
      seti('contour_line_thickness', 1)
      setc('contour_line_colour', 'black')
      setc("contour_highlight",'off')
      setc("contour_label","on")
      
      setc("contour_level_selection_type","level_list")
      wave_level_list = [0.0, 0.5, 1.25, 2.5, 4.0, 6.0, 9.0, 14.0, 999.0]
      set1r('contour_level_list', wave_level_list)

      setc("contour_shade","on")
      setc("contour_shade_method","area_fill")

      setc("contour_shade_colour_method","list")
      wave_level_colors = [60, 40, 100, 190, 240, 280, 300, 0].
       each.inject([]) {|e,v| e << "HSL(#{v},0.5,0.5)"}
      set1c('contour_shade_colour_list', wave_level_colors)

      cont
      coast
    end
  end
end

toto = WW3.new
toto.fetch_grib_file
toto.plot_grib_file

